common/src/Renderer/EntityDecalRenderer.cpp

```diff
@@ -0,0 +1,438 @@
+    auto* currentDest = indexDest;
+    for (const auto& i : indices)
+    {
+      *(currentDest++) = static_cast<GLuint>(brushVerticesStartIndex + i);
+    }
+    data.faceIndicesKey = indexBlock;
+  }
+
+  data.validated = true;
+}
+
+using Vertex = Renderer::GLVertexTypes::P3NT2::Vertex;
+std::vector<Vertex> EntityDecalRenderer::createDecalBrushFace(
+  const Model::EntityNode* entityNode,
+  const Model::BrushNode* brush,
+  const Model::BrushFace& face,
+  const Assets::Texture* texture)
+{
+  assert(texture != nullptr);
+
+  const auto textureWidth = static_cast<float>(texture->width());
+  const auto textureHeight = static_cast<float>(texture->height());
+  const auto textureSize = vm::vec2f{textureWidth, textureHeight};
+  const auto textureName = texture->name();
+
+  // copy the face properties, used to calculate the decal size and texture coords
+  auto attrs = Model::BrushFaceAttributes{textureName, face.attributes()};
+  auto texCoords = face.texCoordSystem().clone();
+  const auto tex = texCoords.get();
+
+  // create the geometry for the decal
+  const auto plane = face.boundary();
+  const auto origin = entityNode->physicalBounds().center();
+  const auto center = plane.project_point(origin);
+
+  // re-project the vertices in case the texture axes are not on the face plane
+  const auto xShift = tex->xAxis() * static_cast<double>(attrs.xScale())
+                      * static_cast<double>(textureWidth) / 2.0;
+  const auto yShift = tex->yAxis() * static_cast<double>(attrs.yScale())
+                      * static_cast<double>(textureHeight) / 2.0;
+
+  // we want to shift every vertex by just a little bit to avoid z-fighting
+  const auto offset = plane.normal * 0.1;
+
+  // start with a rectangle
+  auto verts = std::vector{
+    plane.project_point(center + xShift - yShift) + offset, // Bottom Right
+    plane.project_point(center + xShift + yShift) + offset, // Top Right
+    plane.project_point(center - xShift + yShift) + offset, // Top Left
+    plane.project_point(center - xShift - yShift) + offset  // Bottom Left
+  };
+
+  // because the texture axes don't have to align to the face, we might have a reversed
+  // face here. if so, reverse the points to get a valid face for the plane
+  // todo: is there a better way to do this?
+  const auto [_, vertPlane] = vm::from_points(verts[0], verts[1], verts[2]);
+  if (!vm::is_equal(plane.normal, vertPlane.normal, 0.01))
+  {
+    std::reverse(std::begin(verts), std::end(verts));
+  }
+
+  // calculate the texture offset based on the first vertex location
+  const auto vtx = verts[0];
+  const auto xOffs = -vm::dot(vtx, tex->xAxis()) / attrs.xScale();
+  const auto yOffs = -vm::dot(vtx, tex->yAxis()) / attrs.yScale();
+  attrs.setXOffset(static_cast<float>(xOffs));
+  attrs.setYOffset(static_cast<float>(yOffs));
+
+  // clip the decal geometry against every other plane in the brush
+  for (const auto& f : brush->brush().faces())
+  {
+    if (f == face)
```

Comment:
I think you should be able to compare the addresses here - that's a lot faster.

Reply:
What would be the proper comparison syntax to use here? I thought the `&` represented a reference type, so I thought I was doing the right thing already.

Reply:
& in front of an expression evaluates to the address of that expression (or rather, the address of the value that the expression evaluates to). Comparing addresses is fast because it's just two numbers, essentially.

The `&` you are referring to specifies that we are taking `f` by reference. References are like pointers, but they cannot be reassigned to (the always point to the same object) and they are used like a value, syntactically. That's why you use the `.` operator to access the members of a referenced value, as opposed to the the `->` operator that you use on pointers.

What you are doing in your code is invoke `bool operator==(const BrushFace& lhs, const BrushFace& rhs)`. This operator is generated by [this statement](https://github.com/TrenchBroom/TrenchBroom/blob/fa401dac439ffc272e8a1fe9bbe55af9521617f8/common/src/Model/BrushFace.h#L115C3-L115C19), and it just recursively compares the face's members - which is obviously more costly than comparing two addresses.

C++ is pretty confusing in this regard, but it becomes a bit clearer if you just treat everything like a value. A pointer is a value (you can assign to it, you can modify it, and what's special about it is that you can dereference it). Values and value semantics are the core of C++ - and much of the modernization of C++ since C++11 has been about getting away from pointers and using values and value semantics. Values are cool because you can reason about them more easily, and value semantics make it more easy to parallelize algorithms, because most of the time you are copying and moving them around.

But sometimes you want to reference a value without copying or moving it - that's where references come in. You can use a reference to give a function access to a value without copying or moving the value into the function. The caller keeps the value for itself, but the called function can still access it.

I hope this didn't confuse you more. It's pretty difficult to explain because this is so different from a lot of other languages. C# and Java use reference semantics, for example - they are quite different. I know both of them can also use value semantics now, but their default was always that everything gets passed by reference. But in C++, everything is passed by value per default. When you're passing a pointer, you're just passing a value, but one that happens to be the address of another value. References are special in C++, hence you need special syntax to use them - the trailing `&` after a type expression.

Reply:
Thanks for the detailed explanation - coming from C#, the copy/move semantics in C++ is pretty foreign to me so I'm learning as I go along. Modern C++ is a lot different to the old style C-with-classes that I'm more familiar with.